### ISSUE TITLE
Consider improving the keyboard navigation for the tab list in the Crane app

### ISSUE THREAD

**Issue by vick08** [flutter/flutter#82556]
Presently, the user has to press the TAB key to navigate between the tabs in the Crane app (Fly, Sleep and Eat). Additionally, when any of the tabs is selected, the selection state is not reported to the screen reader and other assistive technology.

Steps to reproduce:
1. Launch VoiceOver screen reader on Mac OS (Command+POWER button three times or ask Siri).
2. Using Chrome or Safari navigate to gallery.flutter.dev, then tab around and choose the "Crane" link, press the ENTER key to go inside the app.
3. Navigate to the list of tabs using the TAB key, then press the ENTER key to select one of the tabs. Observe the behavior and spoken feedback.

Expected results:
According to [ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/#tabpanel)
1. The state of the currently-selected tab should be reflected so that screen reader users can perceive the it.
2. Left and right arrow keys should navigate between the tabs.

Observed results:
None of the above happens.

**Comment by yjbanov** [comment]
Update: it looks like the framework sets a tap handler on the tabs, but it doesn't specify a semantic role, such as button. So the web engine doesn't apply aria-role=button. I'm not sure if the bug is in the framework (perhaps it should specify the role), or in the engine (it should default to a role, such as button, when onTap is present but no role is specified). Will chat with the framework team tomorrow and find out.

**Comment by goderbauer** [comment]
I believe the framework is not marking Tabs as buttons because Android and iOS natively do not give Tabs these roles. If necessary, we could introduce another Role (e.g. `Tab`) that the web engine can translate to the button role and iOS/Android can drop on the floor. Or does the web need to give anything that's tappable (as denoted by an onTap handler) a role of button?

**Comment by yjbanov** [comment]
I just checked, setting [ARIA tab role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Tab_Role) fixes the issue, so introducing a tab role might make sense. If we do that, we should also make sure that `selected` property is set, which we'd translate to `aria-selected="true"`. It is also recommended that we use `aria-controls` to mark the tab view that's being controlled by the tab. However, even without it there's clear improvement.

Having said that, this may not solve the general issue of semantic nodes that have tap action but no corresponding tappable role. Should we consider this a bug in the widget? Or should the engine default to something, e.g. "button"?

**Comment by goderbauer** [comment]
Some of our platforms (notable iOS and Android) are totally fine with having something be tappable without a role associated to it, so making it a bug on the widget level to not have a role seems inappropriate. I don't know enough about a11y services on the web, though. If the web engine would fallback to marking something as a button by default if it is tappable, would that degrade the a11y experience there?

**Comment by yjbanov** [comment]
I think it would be an improvement if we defaulted to button. Otherwise, the element loses its ability to receive click events at all. On the other hand, there's nothing to incentivize the developer to annotate the semantics tree correctly. Perhaps we could add a `--debug` assertion that forces the node to have a role if a tap action is present. I'm not sure how often developers test in debug mode with semantics enabled, so it may not be much of an incentive in practice. Perhaps in `--debug` mode we could always enable semantics?

Not sure what the optimal amount of yak shave is for this particular issue. I think I'd be satisfied with either or both of the following:

- Introduce `isTab` and have the `Tab` widget declare itself to be a tab.
- Fallback to "button" ARIA role if no other tappable role is specified. If we do this, we'd also need to choose a good precedence strategy. For example, what if something is tappable and has a role, but ARIA does not treat the role as tappable? Which role should take precedence? My gut feeling is that we should still fallback to button in that case, because losing interactivity is probably more severe than losing part of the description.
